docs/configuration.rst
.. _configuration:

Configuration
=============

The configuration is stored in the ``config.ini`` file.

+-------------------------+-------------------+-----------------------+----------------------------------------------------+
| Section                 | Setting           | Default value         | Note                                               |
+=========================+===================+=======================+====================================================+
| ``local-control-plane`` | ``host``          | 0.0.0.0               | IP address to accept requests.                     |
+-------------------------+-------------------+-----------------------+----------------------------------------------------+
| ``local-control-plane`` | ``port``          | 4000                  | TCP port to accept requests.                       |
+-------------------------+-------------------+-----------------------+----------------------------------------------------+
| ``local-control-plane`` | ``https``         | false                 | Accept only HTTPS requests.                        |
+-------------------------+-------------------+-----------------------+----------------------------------------------------+
| ``auth``                | ``enabled``       | true                  | Enable JWT authentication.                         |
+-------------------------+-------------------+-----------------------+----------------------------------------------------+
| ``auth``                | ``header-prefix`` | ASTRID                | Header prefix for JWT authentication.              |
+-------------------------+-------------------+-----------------------+----------------------------------------------------+
| ``auth``                | ``secret-key``    | astrid-secret-key     | Secret key for JWT authentication.                 |
+-------------------------+-------------------+-----------------------+----------------------------------------------------+
| ``elastic-apm``         | ``enabled``       | false                 | Enable Elastic APM integration.                    |
+-------------------------+-------------------+-----------------------+----------------------------------------------------+
| ``elastic-apm``         | ``server``        | http://localhost:8200 | Elastic APM server.                                |
+-------------------------+-------------------+-----------------------+----------------------------------------------------+
| ``polycube``            | ``host``          | localhost             | IP address to contact the polycube installation.   |
+-------------------------+-------------------+-----------------------+----------------------------------------------------+
| ``polycube``            | ``port``          | 9000                  | Port address to contact the polycube installation. |
+-------------------------+-------------------+-----------------------+----------------------------------------------------+
| ``polycube``            | ``timeout``       | 20s                   | Timeout for the connection to polycube.            |
+-------------------------+-------------------+-----------------------+----------------------------------------------------+
| ``log``                 | ``config``        | log.yaml              | Path of the log configuration file.                |
+-------------------------+-------------------+-----------------------+----------------------------------------------------+


CHANGELOG.md
# Changelog

## 1.0.0

- Initial release.

## 1.0.1

- Force actions expects "args" to be an array.
- Avoid multiple values for keyword argument during dict merge.
- Added marshmallow in requirements.txt.

## 1.0.2

- Improved execution of action as daemon.
- Accept list of strings for eBPF-Program code.
- Stdout and stderr response format: plain, lines, and JSON.

## 1.0.3

- Minor fixes.

## 1.0.4

- Added option to force the use of HTTPS instead of HTTP.
- Add JWT Authentication.
- Enhanced Log configuration.


reader/config.py
from configparser import BasicInterpolation as Basic_Interpolation
from configparser import ConfigParser as Config_Parser
from os import path
from pathlib import Path


class Config_Reader:
    path = Path(__file__).parent / '../config.ini'

    def __init__(self):
        self.cr = Config_Parser(interpolation=Config_Reader.Env_Interpolation())

    def read(self):
        self.cr.read(self.path.resolve())

        self.lcp_host = self.cr.get('local-control-plane', 'host', fallback='0.0.0.0')
        self.lcp_port = self.cr.get('local-control-plane', 'port', fallback=4000)
        self.lcp_https = self.cr.getboolean('local-control-plane', 'https', fallback=False)

        self.auth = self.cr.getboolean('auth', 'enabled', fallback=True)
        self.auth_header_prefix = self.cr.get('auth', 'header-prefix', fallback='ASTRID')
        self.auth_secret_key = self.cr.get('auth', 'secret-key', fallback='astrid-secret-key')

        self.elastic_apm_enabled = self.cr.getboolean('elastic-apm', 'enabled', fallback=False)
        self.elastic_apm_server = self.cr.get('elastic-apm', 'server', fallback='http://localhost:8200')

        self.polycube_host = self.cr.get('polycube', 'host', fallback='localhost')
        self.polycube_port = self.cr.get('polycube', 'port', fallback=9000)
        self.polycube_timeout = self.cr.get('polycube', 'timeout', fallback='20s')

        self.log_config = self.cr.get('log', 'config', fallback='log.yaml')

    def write(self, db):
        # FIXME is it necessary?
        self.cr.set('local-control-plane', 'port', db.port)

        with self.path.open('w') as f:
            self.cr.write(f)

    class Env_Interpolation(Basic_Interpolation):
        """Interpolation which expands environment variables in values."""

        def before_get(self, parser, section, option, value, defaults):
            """Execute before getting the value.

            :param self: class instance
            :param parser: configparser instance
            :param section: section value
            :param option: option value
            :param value: current value
            :param defaults: default values
            :returns: value with expanded variables
            """
            return path.expandvars(value)
